Add Documentation For Registering Custom Type Converter (#8340)
dotnet/winforms: docs/serialization/register-custom-type-converter.md

# Manually Registering a TypeConverter to a Class

> Typically, a type converter is bound to a class by adding the [TypeConverterAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.typeconverterattribute?view=net-7.0) to the class. 

Reviewer: https://review.learn.microsoft.com/help/platform/links-how-to?branch=main#locale-codes-in-external-links
```suggestion
Typically, a type converter is bound to a class by adding the [TypeConverterAttribute](https://learn.microsoft.com/dotnet/api/system.componentmodel.typeconverterattribute) to the class. 
```

Your inherited class will be added to the `TypeDescriptor`'s list of providers.
We will want to override `GetTypeDescriptor()` to return `MyCustomTypeConverterDescriptor`.
```c#
public class MyTypeDescriptionProvider : TypeDescriptionProvider
    {
        private TypeConverter _myConverter;

        public MyTypeDescriptionProvider(TypeDescriptionProvider parent, TypeConverter converter) : base(parent)
            => _myConverter = converter;

        public override ICustomTypeDescriptor GetTypeDescriptor(
            [DynamicallyAccessedMembers((DynamicallyAccessedMemberTypes)(-1))] Type objectType,
            object instance) => new MyCustomTypeConverterDescriptor(base.GetTypeDescriptor(objectType, instance), _myConverter);
    }
```
Note: The parent `TypeDescriptionProvider` to be passed to the constructor can be grabbed by calling `TypeDescriptor.GetProvider()` and passing in the object/type.

### Registration/Deregistration
Once you have `MyTypeDescriptionProvider` and `MyCustomTypeConverterDescriptor` as outlined above,
register your custom `TypeConverter` to a object/type by calling `TypeDescriptor.AddProvider()` with parameters `MyTypeDescriptionProvider`

> and the object/type you want your custom converter to be associated with. 

Reviewer: Please don't hard wrap sentences, it makes it harder to review and to make suggestions.

```c#
private class MyCustomTypeConverterDescriptor : CustomTypeDescriptor
{
    private TypeConverter _myConverter;

    public MyCustomTypeConverterDescriptor(ICustomTypeDescriptor parent, TypeConverter converter) : base(parent)
        => _myConverter = converter;

    public override TypeConverter GetConverter() => _myConverter;
}
```

`ICustomTypeDescriptor` is accessed through [`TypeDescriptionProvider.GetTypeDescriptor()`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptionprovider.gettypedescriptor). This method needs to be overridden to provide your newly created `ICustomTypeDescriptor` which provides your custom type converter. Have another class inherit from `TypeDescriptionProvider` to preserve the default behavior and override `GetTypeDescriptor()` to provide `MyCustomTypeConverterDescriptor`.

```c#

> public class MyTypeDescriptionProvider : TypeDescriptionProvider

Reviewer 2: I'd consider being more formal as this is usable as is. `CustomConverterProvider` perhaps. Likewise for the class above `CustomConverterDescriptor`. (That way it can be copy / pasted.)

{
    private TypeConverter _myConverter;

    public MyTypeDescriptionProvider(TypeDescriptionProvider parent, TypeConverter converter) : base(parent)
        => _myConverter = converter;

    public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object instance) 
        => new MyCustomTypeConverterDescriptor(base.GetTypeDescriptor(objectType, instance), _myConverter);

>     // MyCustomTypeConverterDescriptor code...

Reviewer 2: ```suggestion
    // Consider nesting MyCustomTypeConverterDescriptor here
```

### Registration/Deregistration
Once you have `MyTypeDescriptionProvider` and `MyCustomTypeConverterDescriptor` as outlined above, finish up registering your custom `TypeConverter` to a class by calling [`TypeDescriptor.AddProvider()`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptor.addprovider) with parameters `MyTypeDescriptionProvider` and the class you want your custom converter to be associated with. 
```c#
TypeDescriptionProvider parentProvider = TypeDescriptor.GetProvider(classType);
MyCustomTypeConverterDescriptor newProvider = new(parentProvider, myConverter);
TypeDescriptor.AddProvider(newProvider, classType);
```

At this point, calling [`TypeDescriptor.GetConverter()`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptor.getconverter) with the same class that was passed to `TypeDescriptor.AddProvider()` will now return your custom type converter! 

> If you want this type converter to be temporarily registered to the class, do not forget to call [`TypeDescriptor.RemoveProvider()`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptor.removeprovider) when finished.

Reviewer 2: ```suggestion
To restore the previous behavior call [`TypeDescriptor.RemoveProvider()`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptor.removeprovider).
```

# Manually Registering a TypeConverter to a Class

> A type converter is bound to a class by adding the [TypeConverterAttribute](https://learn.microsoft.com/dotnet/api/system.componentmodel.typeconverterattribute) to the class. However, there are some scenarios where it is not possible to add the attribute to the class. e.g. user may not own the class, but there is no type converter registered to that class for serialization. This document outlines how to manually register a [TypeConverter](https://learn.microsoft.com/dotnet/api/system.componentmodel.typeconverter) to a class for these scenarios.

Reviewer: May worth formatting all API like this:
```suggestion
A type converter is bound to a class by adding the [`TypeConverterAttribute`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typeconverterattribute) to the class. However, there are some scenarios where it is not possible to add the attribute to the class. e.g., user may not own the class, but there is no type converter registered to that class for serialization. This document outlines how to manually register a [`TypeConverter`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typeconverter) to a class for these scenarios.
```

        => _myConverter = converter;

    public override TypeConverter GetConverter() => _myConverter;
}

```
`ICustomTypeDescriptor` is accessed through [TypeDescriptionProvider.GetTypeDescriptor](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptionprovider.gettypedescriptor). `TypeDescriptionProvider` is an existing abstract class that already has default behavior. `GetTypeDescriptor()` is the only method is important for registering your new `ICustomTypeDescriptor` and thus registering your custom type converter. Have another class inherit from `TypeDescriptionProvider` to get its default behavior and override `GetTypeDescriptor()` to provide `MyCustomTypeConverterDescriptor`.

```c#
public class MyTypeDescriptionProvider : TypeDescriptionProvider
{
    private TypeConverter _myConverter;

    public MyTypeDescriptionProvider(TypeDescriptionProvider parent, TypeConverter converter) : base(parent)
        => _myConverter = converter;

    public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object instance) 
        => new MyCustomTypeConverterDescriptor(base.GetTypeDescriptor(objectType, instance), _myConverter);
}
```

> Note: The parent `TypeDescriptionProvider` to be passed to the constructor can be grabbed by calling [TypeDescriptor.GetProvider()]() and passing in the class that is intended to be associated to your custom type converter.

Reviewer: I suppose you missed this one: ![image](https://user-images.githubusercontent.com/4403806/206066457-4b36e590-da03-44ae-9af5-cb4c89b67df7.png)

```mermaid
sequenceDiagram
TypeDescriptor-->MyTypeDescriptionProvider_Inherits_TypeDescriptionProvider: GetConverter()
Note over TypeDescriptor,MyTypeDescriptionProvider_Inherits_TypeDescriptionProvider: Must call TypeDescriptor.AddProvider() with a given class first
Note right of MyTypeDescriptionProvider_Inherits_TypeDescriptionProvider: Call TypeDescriptor.RemoveProvider() when no longer want to associate your custom type converter with a given class
MyTypeDescriptionProvider_Inherits_TypeDescriptionProvider-->MyCustomTypeDescriptor_Inherits_CustomTypeDescriptor: GetTypeDescriptor()
MyCustomTypeDescriptor-->MyTypeConverter_Inherits_TypeConverter: GetConverter()

> ```

Reviewer: Nice!

I'm using https://sequencediagram.org/ and then exporting images from it.

Author: Thank you! Always coming in with great tips 🙏

```c#
TypeDescriptor.RemoveProvider(newProvider, type);
```

## Additional Information

### TypeDescriptor.GetConverter() Details
This is a static method on [`TypeDescriptor`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptor) that will run through its private list of [`TypeDescriptionProviders`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptionprovider) and get the provider associated with the type. Once the right provider is found, it will then call [`TypeDescriptionProvider.GetTypeDescriptor()`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typedescriptionprovider.gettypedescriptor) to get an [ICustomTypeDescriptor](https://learn.microsoft.com/dotnet/api/system.componentmodel.icustomtypedescriptor) and call [`ICustomTypeDescriptor.GetConverter()`](https://learn.microsoft.com/dotnet/api/system.componentmodel.icustomtypedescriptor.getconverter) to grab the [`TypeConverter`](https://learn.microsoft.com/dotnet/api/system.componentmodel.typeconverter). 

> ```mermaid

Reviewer 2: Love this!
